src/tribler/core/components/restapi/rest/utils.py

```diff
@@ -66,3 +88,43 @@ def fix_unicode_array(arr):
             new_arr.append(item)
 
     return new_arr
+
+
+def shorten(s: Optional[str], width: int = 50, placeholder='[...]', cut_at_the_end: bool = True):
+    """ Shorten a string to a given width.
+        Example: shorten('hello world', 5) -> 'hello[...]'
+    """
+    if s and len(s) > width:
+        return f'{s[:width]}{placeholder}' if cut_at_the_end else f'{placeholder}{s[len(s) - width:]}'
+    return s
+
+
+def _format_frames(frame: Optional[FrameType], file_width: int = 50, value_width: int = 100) -> Iterator[str]:
+    """ Format a stack trace."""
+    while frame:
+        filename = shorten(frame.f_code.co_filename, width=file_width, cut_at_the_end=False)
+        header = f"{filename}:{frame.f_lineno} {frame.f_code.co_name}"
+        local = ''
+        for key, value in list(frame.f_locals.items()):
+            value = shorten(repr(value), width=value_width)
+            local += f'\n\t{key} = {value}'
+        frame = frame.f_back
+        yield header + local
+
+
+def get_threads_info() -> List[Dict]:
+    """
+    Return information about available threads.
+    """
+
+    result = []
+    with switch_interval(ONE_SECOND):
```

Comment:
It is important to also use `with switch_interval` in `tribler.coro.utilities.slow_coro_detection.main_thread_stack_tracking._get_main_thread_stack_info`; otherwise, it does not protect from the race condition.

Reply:
Like this?

```python

def _get_main_thread_stack_info() -> List[StackFrameInfo]:
    """
    Quickly copies necessary information from the main thread stack, so it is possible later to format a usual
    traceback in a separate thread.

    The function temporarily changes the interpreter’s thread switch interval to prevent thread switch during
    the stack copying. It is a lighter analogue of holding the GIL (Global Interpreter Lock).
    """
    with switch_interval(SWITCH_INTERVAL):
        stack_info = []
        for frame, start_time in _main_thread_stack:
            func_name = frame.f_code.co_name

            if func_name == 'profile_wrapper':
                # We have profiler enabled for this function. Do not show nested frames (if present). Instead of
                # the name `profile_wrapper`, let's show the original function name to make the traceback clearer
                original_func: FunctionType = frame.f_locals.get('func')
                if original_func is not None:
                    stack_info.append(StackFrameInfo(
                        func_name=original_func.__name__,
                        file_name=original_func.__code__.co_filename,
                        start_time=start_time,
                        is_under_profiling=True
                    ))
                    break

            stack_info.append(StackFrameInfo(
                func_name=func_name,
                file_name=frame.f_code.co_filename,
                line_number=frame.f_lineno,
                start_time=start_time
            ))
            
    return stack_info
```

Reply:
Exactly!